File: TradingTracker.java
import javax.swing.*;
import javax.swing.border.EmptyBorder;
import javax.swing.table.AbstractTableModel;
import java.awt.*;
import java.awt.event.*;
import java.util.ArrayList;
import java.util.Random;

public class TradingTracker extends JFrame {

    private TransactionTracker tracker;
    private PortfolioManager portfolio;
    private ArrayList<Trade> allTrades;

    private JTextField symbolField;
    private JTextField priceField;
    private JTextField volumeField;
    private JTable tradeTable;
    private TradeTableModel tableModel;
    private JLabel bestTradeLabel;
    private JLabel worstTradeLabel;

    private Random random;
    private javax.swing.Timer priceUpdateTimer;

    public TradingTracker() {
        tracker = new TransactionTracker();
        portfolio = new PortfolioManager();
        allTrades = new ArrayList<>();
        random = new Random();

        setTitle("Stock Market Trading Tracker");
        setSize(800, 600);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setLocationRelativeTo(null); // Center the window.

        // Main container - BorderLayout.
        JPanel mainPanel = new JPanel(new BorderLayout(10, 10));
        mainPanel.setBorder(new EmptyBorder(10, 10, 10, 10));
        add(mainPanel);

        // Top 
        JPanel topPanel = new JPanel();
        topPanel.setLayout(new BoxLayout(topPanel, BoxLayout.Y_AXIS));
        mainPanel.add(topPanel, BorderLayout.NORTH);

        // Trade details panel  - GridLayout.
        JPanel detailsPanel = new JPanel(new GridLayout(1, 6, 5, 5));
        detailsPanel.setBorder(BorderFactory.createTitledBorder("Trade Details"));
        topPanel.add(detailsPanel);

        detailsPanel.add(new JLabel("Stock Symbol:"));
        symbolField = new JTextField();
        detailsPanel.add(symbolField);

        detailsPanel.add(new JLabel("Price:"));
        priceField = new JTextField("100");
        detailsPanel.add(priceField);

        detailsPanel.add(new JLabel("Volume:"));
        volumeField = new JTextField("10");
        detailsPanel.add(volumeField);

        // Buttons panel- FlowLayout centered.
        JPanel buttonPanel = new JPanel(new FlowLayout(FlowLayout.CENTER, 15, 5));
        topPanel.add(buttonPanel);
        buttonPanel.setBorder(BorderFactory.createEmptyBorder(5, 0, 5, 0));

        JButton addButton = new JButton("Add Trade");
        JButton sellButton = new JButton("Sell Stock");
        JButton updateAllButton = new JButton("Update All Prices");
        JButton updateSelectedButton = new JButton("Update Selected Stock");
        buttonPanel.add(addButton);
        buttonPanel.add(sellButton);
        buttonPanel.add(updateAllButton);
        buttonPanel.add(updateSelectedButton);

        //   table 
        tableModel = new TradeTableModel(allTrades);
        tradeTable = new JTable(tableModel);
        JScrollPane tableScroll = new JScrollPane(tradeTable);
        tableScroll.setBorder(BorderFactory.createTitledBorder("Trades"));
        mainPanel.add(tableScroll, BorderLayout.CENTER);

        // Summary 
        JPanel summaryPanel = new JPanel(new FlowLayout(FlowLayout.LEFT, 10, 5));
        bestTradeLabel = new JLabel("Best Trade: N/A");
        worstTradeLabel = new JLabel("Worst Trade: N/A");
        summaryPanel.add(bestTradeLabel);
        summaryPanel.add(worstTradeLabel);
        mainPanel.add(summaryPanel, BorderLayout.SOUTH);

        // Button action
        addButton.addActionListener(e -> addTrade());
        sellButton.addActionListener(e -> sellTrade());
        updateAllButton.addActionListener(e -> updateAllPrices());
        updateSelectedButton.addActionListener(e -> updateSelectedStock());

        //  Swing Timer - 10 seconds.
        priceUpdateTimer = new javax.swing.Timer(10000, e -> updateAllPrices());
        priceUpdateTimer.start();

        updateSummary();
    }

    // Buy
    private void addTrade() {
        String symbol = symbolField.getText().trim();
        if (symbol.isEmpty()) {
            JOptionPane.showMessageDialog(this, "Please enter a stock symbol.", "Input Error", JOptionPane.ERROR_MESSAGE);
            return;
        }
        double price;
        int volume;
        try {
            price = Double.parseDouble(priceField.getText());
            volume = Integer.parseInt(volumeField.getText());
        } catch (NumberFormatException ex) {
            JOptionPane.showMessageDialog(this, "Price must be a number and volume must be an integer.", "Input Error", JOptionPane.ERROR_MESSAGE);
            return;
        }
        double timestamp = System.currentTimeMillis() / 1000.0;
        Trade trade = new Trade(timestamp, symbol, price, volume, price, "Buy");
        tracker.addTrade(trade);
        portfolio.addTrade(trade);
        allTrades.add(trade);
        symbolField.setText("");
        tableModel.fireTableDataChanged();
        updateSummary();
    }

    // Sell 
    private void sellTrade() {
        String symbol = JOptionPane.showInputDialog(this, "Enter the stock symbol to sell:");
        if (symbol == null || symbol.trim().isEmpty()) {
            return;
        }
        Trade buyTrade = null;
        for (Trade t : allTrades) {
            if (t.getSymbol().equalsIgnoreCase(symbol.trim()) && t.getTradeType().equals("Buy") && t.getVolume() > 0) {
                buyTrade = t;
                break;
            }
        }
        if (buyTrade == null) {
            JOptionPane.showMessageDialog(this, "No available buy trade found for symbol: " + symbol, "Sell Error", JOptionPane.ERROR_MESSAGE);
            return;
        }
        String volStr = JOptionPane.showInputDialog(this, "Enter number of shares to sell (Available: " + buyTrade.getVolume() + "):");
        if (volStr == null || volStr.trim().isEmpty()) {
            return;
        }
        int sellVolume;
        try {
            sellVolume = Integer.parseInt(volStr);
        } catch (NumberFormatException ex) {
            JOptionPane.showMessageDialog(this, "Volume must be an integer.", "Input Error", JOptionPane.ERROR_MESSAGE);
            return;
        }
        if (sellVolume <= 0 || sellVolume > buyTrade.getVolume()) {
            JOptionPane.showMessageDialog(this, "Invalid sell volume. Must be between 1 and " + buyTrade.getVolume(), "Sell Error", JOptionPane.ERROR_MESSAGE);
            return;
        }
        double currentPrice = buyTrade.getPrice();
        double timestamp = System.currentTimeMillis() / 1000.0;
        Trade sellTrade = new Trade(timestamp, symbol, currentPrice, sellVolume, buyTrade.getOriginalPrice(), "Sell");
        tracker.addTrade(sellTrade);
        portfolio.addTrade(sellTrade);
        allTrades.add(sellTrade);
        buyTrade.reduceVolume(sellVolume);
        tableModel.fireTableDataChanged();
        updateSummary();
    }

    // price update.
    private double simulatePriceUpdate(Trade trade, double maxFluctuation) {
        double factor = 1 + (random.nextDouble() * 2 * maxFluctuation - maxFluctuation);
        return trade.getPrice() * factor;
    }

    // Update all 
    private void updateAllPrices() {
        for (Trade trade : allTrades) {
            if (trade.getTradeType().equals("Buy") && trade.getVolume() > 0) {
                double newPrice = simulatePriceUpdate(trade, 0.05);
                trade.setPrice(newPrice);
            }
        }
        tracker = new TransactionTracker();
        portfolio = new PortfolioManager();
        for (Trade trade : allTrades) {
            tracker.addTrade(trade);
            portfolio.addTrade(trade);
        }
        tableModel.fireTableDataChanged();
        updateSummary();
    }

    // Update for selected trade.
    private void updateSelectedStock() {
        int selectedRow = tradeTable.getSelectedRow();
        if (selectedRow == -1) {
            JOptionPane.showMessageDialog(this, "Please select a trade to update.", "No Selection", JOptionPane.WARNING_MESSAGE);
            return;
        }
        Trade trade = allTrades.get(selectedRow);
        if (!trade.getTradeType().equals("Buy") || trade.getVolume() <= 0) {
            JOptionPane.showMessageDialog(this, "Selected trade is not an active buy trade.", "Update Error", JOptionPane.ERROR_MESSAGE);
            return;
        }
        double newPrice = simulatePriceUpdate(trade, 0.05);
        trade.setPrice(newPrice);
        tracker = new TransactionTracker();
        portfolio = new PortfolioManager();
        for (Trade t : allTrades) {
            tracker.addTrade(t);
            portfolio.addTrade(t);
        }
        tableModel.fireTableDataChanged();
        updateSummary();
    }

    // Update best and worst trade 
    private void updateSummary() {
        Trade best = tracker.getBestTrade();
        Trade worst = tracker.getWorstTrade();
        bestTradeLabel.setText(best != null ? "Best Trade: " + best.getSymbol() + " | Profit: " + String.format("%.2f", best.performanceMetric()) : "Best Trade: N/A");
        worstTradeLabel.setText(worst != null ? "Worst Trade: " + worst.getSymbol() + " | Profit: " + String.format("%.2f", worst.performanceMetric()) : "Worst Trade: N/A");
    }

    public static void main(String[] args) {
        SwingUtilities.invokeLater(() -> {
            TradingTracker trackerApp = new TradingTracker();
            trackerApp.setVisible(true);
        });
    }
}

// Table model to display trades
class TradeTableModel extends AbstractTableModel {

    private final String[] columnNames = {"Timestamp", "Symbol", "Type", "Price", "Volume", "Orig. Price", "Performance"};
    private ArrayList<Trade> trades;

    public TradeTableModel(ArrayList<Trade> trades) {
        this.trades = trades;
    }

    @Override
    public int getRowCount() {
        return trades.size();
    }

    @Override
    public int getColumnCount() {
        return columnNames.length;
    }
    
    @Override
    public String getColumnName(int col) {
        return columnNames[col];
    }
    
    @Override
    public Object getValueAt(int row, int col) {
        Trade trade = trades.get(row);
        switch (col) {
            case 0:
                return String.format("%.2f", trade.getTimestamp());
            case 1:
                return trade.getSymbol();
            case 2:
                return trade.getTradeType();
            case 3:
                return String.format("%.2f", trade.getPrice());
            case 4:
                return trade.getVolume();
            case 5:
                return String.format("%.2f", trade.getOriginalPrice());
            case 6:
                return String.format("%.2f", trade.performanceMetric());
            default:
                return "";
        }
    }
}


class Trade {
    private double timestamp;
    private String symbol;
    private double price;
    private int volume;
    private double originalPrice;
    private String tradeType;

    public Trade(double timestamp, String symbol, double price, int volume, double originalPrice, String tradeType) {
        this.timestamp = timestamp;
        this.symbol = symbol;
        this.price = price;
        this.volume = volume;
        this.originalPrice = originalPrice;
        this.tradeType = tradeType;
    }

    public double performanceMetric() {
        return (price - originalPrice) * volume;
    }

    public double getTimestamp() {
        return timestamp;
    }

    public String getSymbol() {
        return symbol;
    }

    public double getPrice() {
        return price;
    }

    public void setPrice(double price) {
        if (tradeType.equals("Buy")) {
            this.price = price;
        }
    }

    public int getVolume() {
        return volume;
    }

    public void reduceVolume(int amount) {
        this.volume -= amount;
        if (this.volume < 0) {
            this.volume = 0;
        }
    }

    public double getOriginalPrice() {
        return originalPrice;
    }

    public String getTradeType() {
        return tradeType;
    }

    @Override
    public String toString() {
        return "Trade(Symbol: " + symbol +
               ", Type: " + tradeType +
               ", Price: " + String.format("%.2f", price) +
               ", Volume: " + volume +
               ", Orig.Price: " + String.format("%.2f", originalPrice) +
               ", Time: " + String.format("%.2f", timestamp) + ")";
    }
}


class TransactionTracker {
    private ArrayList<Trade> trades;

    public TransactionTracker() {
        trades = new ArrayList<>();
    }

    public void addTrade(Trade trade) {
        trades.add(trade);
    }

    public Trade getBestTrade() {
        if (trades.isEmpty()) return null;
        Trade best = trades.get(0);
        for (Trade trade : trades) {
            if (trade.performanceMetric() > best.performanceMetric()) {
                best = trade;
            }
        }
        return best;
    }

    public Trade getWorstTrade() {
        if (trades.isEmpty()) return null;
        Trade worst = trades.get(0);
        for (Trade trade : trades) {
            if (trade.performanceMetric() < worst.performanceMetric()) {
                worst = trade;
            }
        }
        return worst;
    }
}

class AVLNode {
    double key;
    Trade trade;
    int height;
    AVLNode left;
    AVLNode right;

    public AVLNode(double key, Trade trade) {
        this.key = key;
        this.trade = trade;
        this.height = 1;
    }
}

class PortfolioManager {
    private AVLNode root;

    public PortfolioManager() {
        root = null;
    }

    private int getHeight(AVLNode node) {
        return (node == null) ? 0 : node.height;
    }

    private int getBalance(AVLNode node) {
        return (node == null) ? 0 : getHeight(node.left) - getHeight(node.right);
    }

    private AVLNode rightRotate(AVLNode y) {
        AVLNode x = y.left;
        AVLNode T2 = x.right;
        x.right = y;
        y.left = T2;
        y.height = 1 + Math.max(getHeight(y.left), getHeight(y.right));
        x.height = 1 + Math.max(getHeight(x.left), getHeight(x.right));
        return x;
    }

    private AVLNode leftRotate(AVLNode x) {
        AVLNode y = x.right;
        AVLNode T2 = y.left;
        y.left = x;
        x.right = T2;
        x.height = 1 + Math.max(getHeight(x.left), getHeight(x.right));
        y.height = 1 + Math.max(getHeight(y.left), getHeight(y.right));
        return y;
    }

    private AVLNode insertNode(AVLNode node, double key, Trade trade) {
        if (node == null) {
            return new AVLNode(key, trade);
        }
        if (key < node.key) {
            node.left = insertNode(node.left, key, trade);
        } else {
            node.right = insertNode(node.right, key, trade);
        }
        node.height = 1 + Math.max(getHeight(node.left), getHeight(node.right));
        int balance = getBalance(node);
        if (balance > 1 && key < node.left.key) {
            return rightRotate(node);
        }
        if (balance < -1 && key > node.right.key) {
            return leftRotate(node);
        }
        if (balance > 1 && key > node.left.key) {
            node.left = leftRotate(node.left);
            return rightRotate(node);
        }
        if (balance < -1 && key < node.right.key) {
            node.right = rightRotate(node.right);
            return leftRotate(node);
        }
        return node;
    }

    public void addTrade(Trade trade) {
        root = insertNode(root, trade.getTimestamp(), trade);
    }

    private void inorderTraversal(AVLNode node, ArrayList<Trade> result) {
        if (node != null) {
            inorderTraversal(node.left, result);
            result.add(node.trade);
            inorderTraversal(node.right, result);
        }
    }

    public ArrayList<Trade> getInorder() {
        ArrayList<Trade> result = new ArrayList<>();
        inorderTraversal(root, result);
        return result;
    }
}

File: TradingTracker.py
import tkinter as tk
from tkinter import ttk, messagebox, simpledialog
import random
import time
from datetime import datetime

import matplotlib
matplotlib.use("TkAgg")
from matplotlib.figure import Figure
from matplotlib.backends.backend_tkagg import FigureCanvasTkAgg
import matplotlib.dates as mdates

class Trade:
    def __init__(self, timestamp, symbol, price, volume, original_price, trade_type):
        self.timestamp = timestamp
        self.symbol = symbol
        self.price = price
        self.volume = volume
        self.original_price = original_price
        self.trade_type = trade_type

    def performance_metric(self):
        return (self.price - self.original_price) * self.volume

    def reduce_volume(self, amount):
        self.volume -= amount
        if self.volume < 0:
            self.volume = 0

    def __str__(self):
        return (f"Trade(Symbol: {self.symbol}, Type: {self.trade_type}, Price: {self.price:.2f}, "
                f"Volume: {self.volume}, Orig.Price: {self.original_price:.2f}, Time: {self.timestamp:.2f})")

class Heap:
    def __init__(self, comparator):
        self.data = []
        self.comparator = comparator

    def push(self, item):
        self.data.append(item)
        self._sift_up(len(self.data) - 1)

    def pop(self):
        if not self.data:
            return None
        top_item = self.data[0]
        self.data[0] = self.data[-1]
        self.data.pop()
        self._sift_down(0)
        return top_item

    def peek(self):
        if not self.data:
            return None
        return self.data[0]

    def _sift_up(self, index):
        parent_index = (index - 1) // 2
        while index > 0 and self.comparator(self.data[index], self.data[parent_index]):
            self.data[index], self.data[parent_index] = self.data[parent_index], self.data[index]
            index = parent_index
            parent_index = (index - 1) // 2

    def _sift_down(self, index):
        total_items = len(self.data)
        while True:
            left_index = 2 * index + 1
            right_index = 2 * index + 2
            best_index = index
            if left_index < total_items and self.comparator(self.data[left_index], self.data[best_index]):
                best_index = left_index
            if right_index < total_items and self.comparator(self.data[right_index], self.data[best_index]):
                best_index = right_index
            if best_index == index:
                break
            self.data[index], self.data[best_index] = self.data[best_index], self.data[index]
            index = best_index

class TransactionTracker:
    def __init__(self):
        self.trades = []
        self.best_heap = Heap(lambda first_trade, second_trade: first_trade.performance_metric() > second_trade.performance_metric())
        self.worst_heap = Heap(lambda first_trade, second_trade: first_trade.performance_metric() < second_trade.performance_metric())

    def add_trade(self, trade):
        self.trades.append(trade)
        self.best_heap.push(trade)
        self.worst_heap.push(trade)

    def get_best_trade(self):
        return self.best_heap.peek()

    def get_worst_trade(self):
        return self.worst_heap.peek()

class AVLNode:
    def __init__(self, key, trade):
        self.key = key
        self.trade = trade
        self.height = 1
        self.left = None
        self.right = None

class PortfolioManager:
    def __init__(self):
        self.root = None

    def get_height(self, node):
        return node.height if node else 0

    def get_balance(self, node):
        return self.get_height(node.left) - self.get_height(node.right) if node else 0

    def right_rotate(self, node_y):
        node_x = node_y.left
        subtree_T2 = node_x.right
        node_x.right = node_y
        node_y.left = subtree_T2
        node_y.height = 1 + max(self.get_height(node_y.left), self.get_height(node_y.right))
        node_x.height = 1 + max(self.get_height(node_x.left), self.get_height(node_x.right))
        return node_x

    def left_rotate(self, node_x):
        node_y = node_x.right
        subtree_T2 = node_y.left
        node_y.left = node_x
        node_x.right = subtree_T2
        node_x.height = 1 + max(self.get_height(node_x.left), self.get_height(node_x.right))
        node_y.height = 1 + max(self.get_height(node_y.left), self.get_height(node_y.right))
        return node_y

    def insert_node(self, current_node, key, trade):
        if not current_node:
            return AVLNode(key, trade)
        if key < current_node.key:
            current_node.left = self.insert_node(current_node.left, key, trade)
        else:
            current_node.right = self.insert_node(current_node.right, key, trade)
        current_node.height = 1 + max(self.get_height(current_node.left), self.get_height(current_node.right))
        balance_factor = self.get_balance(current_node)
        if balance_factor > 1 and key < current_node.left.key:
            return self.right_rotate(current_node)
        if balance_factor < -1 and key > current_node.right.key:
            return self.left_rotate(current_node)
        if balance_factor > 1 and key > current_node.left.key:
            current_node.left = self.left_rotate(current_node.left)
            return self.right_rotate(current_node)
        if balance_factor < -1 and key < current_node.right.key:
            current_node.right = self.right_rotate(current_node.right)
            return self.left_rotate(current_node)
        return current_node

    def add_trade(self, trade):
        self.root = self.insert_node(self.root, trade.timestamp, trade)

    def inorder_traversal(self, node, result):
        if node:
            self.inorder_traversal(node.left, result)
            result.append(node.trade)
            self.inorder_traversal(node.right, result)

    def get_inorder(self):
        result = []
        self.inorder_traversal(self.root, result)
        return result

class TradingTracker(tk.Tk):
    CANDLE_PERIOD = 10

    def __init__(self):
        super().__init__()
        self.attributes("-fullscreen", True)
        self.transaction_tracker = TransactionTracker()
        self.portfolio_manager = PortfolioManager()
        self.all_trades = []
        self.random_generator = random.Random()
        self.wallet = 10000.0
        self.stock_history = {}
        self.current_symbol = None
        main_panel = tk.Frame(self, padx=10, pady=10)
        main_panel.pack(fill=tk.BOTH, expand=True)
        top_panel = tk.Frame(main_panel)
        top_panel.pack(side=tk.TOP, fill=tk.X)
        details_panel = tk.LabelFrame(top_panel, text="Trade Details", padx=5, pady=5)
        details_panel.pack(fill=tk.X, pady=(0, 5))
        tk.Label(details_panel, text="Stock Symbol:").grid(row=0, column=0, padx=5, pady=5, sticky="e")
        self.symbol_field = tk.Entry(details_panel)
        self.symbol_field.grid(row=0, column=1, padx=5, pady=5, sticky="w")
        tk.Label(details_panel, text="Price:").grid(row=0, column=2, padx=5, pady=5, sticky="e")
        self.price_field = tk.Entry(details_panel)
        self.price_field.insert(0, "100")
        self.price_field.grid(row=0, column=3, padx=5, pady=5, sticky="w")
        tk.Label(details_panel, text="Volume:").grid(row=0, column=4, padx=5, pady=5, sticky="e")
        self.volume_field = tk.Entry(details_panel)
        self.volume_field.insert(0, "10")
        self.volume_field.grid(row=0, column=5, padx=5, pady=5, sticky="w")
        button_panel = tk.Frame(top_panel)
        button_panel.pack(fill=tk.X, pady=(5, 5))
        button_options = {'bd': 2, 'relief': 'groove', 'font': ("Helvetica", 10, "bold"), 'width': 16, 'padx': 5, 'pady': 5}
        self.add_button = tk.Button(button_panel, text="Add Trade", command=self.add_trade, bg="lightgreen", **button_options)
        self.sell_button = tk.Button(button_panel, text="Sell Stock", command=self.sell_trade, bg="tomato", **button_options)
        self.update_all_button = tk.Button(button_panel, text="Update All Prices", command=self.update_all_prices, bg="lightblue", **button_options)
        self.update_selected_button = tk.Button(button_panel, text="Update Selected Stock", command=self.update_selected_stock, bg="orange", **button_options)
        self.add_button.pack(side=tk.LEFT, expand=True, padx=15, pady=5)
        self.sell_button.pack(side=tk.LEFT, expand=True, padx=15, pady=5)
        self.update_all_button.pack(side=tk.LEFT, expand=True, padx=15, pady=5)
        self.update_selected_button.pack(side=tk.LEFT, expand=True, padx=15, pady=5)
        center_pane = ttk.PanedWindow(main_panel, orient=tk.HORIZONTAL)
        center_pane.pack(fill=tk.BOTH, expand=True)
        left_frame = tk.Frame(center_pane)
        center_pane.add(left_frame, weight=1)
        table_frame = tk.LabelFrame(left_frame, text="Trades", padx=5, pady=5)
        table_frame.pack(fill=tk.BOTH, expand=True)
        columns = ("Timestamp", "Symbol", "Type", "Price", "Volume", "Orig. Price", "Performance")
        self.trade_table = ttk.Treeview(table_frame, columns=columns, show="headings", selectmode="browse")
        for column in columns:
            self.trade_table.heading(column, text=column)
            self.trade_table.column(column, anchor="center")
        self.trade_table.pack(fill=tk.BOTH, expand=True)
        vertical_scrollbar = ttk.Scrollbar(table_frame, orient="vertical", command=self.trade_table.yview)
        self.trade_table.configure(yscrollcommand=vertical_scrollbar.set)
        vertical_scrollbar.pack(side=tk.RIGHT, fill=tk.Y)
        right_frame = tk.Frame(center_pane)
        center_pane.add(right_frame, weight=1)
        stock_list_frame = tk.LabelFrame(right_frame, text="Stocks", padx=5, pady=5)
        stock_list_frame.pack(fill=tk.X)
        self.stock_listbox = tk.Listbox(stock_list_frame, height=4)
        self.stock_listbox.pack(fill=tk.X, padx=5, pady=5)
        self.stock_listbox.bind("<<ListboxSelect>>", self.on_stock_select)
        self.selected_stock_label = tk.Label(stock_list_frame, text="Selected Stock: None", font=("Helvetica", 10, "bold"))
        self.selected_stock_label.pack(pady=(0, 5))
        graph_frame = tk.LabelFrame(right_frame, text="Price vs Time", padx=5, pady=5)
        graph_frame.pack(fill=tk.BOTH, expand=True, pady=(5, 0))
        self.figure = Figure(figsize=(4, 3), dpi=100)
        self.ax = self.figure.add_subplot(111)
        self.ax.set_title("Stock Price vs Time")
        self.ax.set_xlabel("Time")
        self.ax.set_ylabel("Price")
        self.canvas = FigureCanvasTkAgg(self.figure, master=graph_frame)
        self.canvas.get_tk_widget().pack(fill=tk.BOTH, expand=True)
        summary_panel = tk.Frame(main_panel)
        summary_panel.pack(side=tk.BOTTOM, fill=tk.X, pady=(5, 0))
        self.best_trade_label = tk.Label(summary_panel, text="Best Trade: N/A")
        self.worst_trade_label = tk.Label(summary_panel, text="Worst Trade: N/A")
        self.wallet_label = tk.Label(summary_panel, text="Wallet: $10000.00")
        self.best_trade_label.pack(side=tk.LEFT, padx=10)
        self.worst_trade_label.pack(side=tk.LEFT, padx=10)
        self.wallet_label.pack(side=tk.LEFT, padx=10)
        self.price_update_timer()
        self.update_summary()
        self.update_stock_list()

    def price_update_timer(self):
        self.update_all_prices()
        if self.current_symbol:
            self.plot_stock_history(self.current_symbol)
        self.after(5000, self.price_update_timer)

    def add_trade(self):
        symbol = self.symbol_field.get().strip()
        if not symbol:
            messagebox.showerror("Input Error", "Please enter a stock symbol.")
            return
        try:
            price = float(self.price_field.get())
            volume = int(self.volume_field.get())
        except ValueError:
            messagebox.showerror("Input Error", "Price must be a number and volume must be an integer.")
            return
        trade_timestamp = time.time()
        new_trade = Trade(trade_timestamp, symbol, price, volume, price, "Buy")
        self.transaction_tracker.add_trade(new_trade)
        self.portfolio_manager.add_trade(new_trade)
        self.all_trades.append(new_trade)
        cost_of_purchase = price * volume
        self.wallet -= cost_of_purchase
        self.update_stock_history(symbol, price, trade_timestamp)
        self.symbol_field.delete(0, tk.END)
        self.refresh_table()
        self.update_summary()
        self.update_stock_list()

    def sell_trade(self):
        symbol = simpledialog.askstring("Sell Stock", "Enter the stock symbol to sell:")
        if symbol is None or not symbol.strip():
            return
        symbol = symbol.strip()
        buy_trade = None
        for trade in self.all_trades:
            if trade.symbol.lower() == symbol.lower() and trade.trade_type == "Buy" and trade.volume > 0:
                buy_trade = trade
                break
        if buy_trade is None:
            messagebox.showerror("Sell Error", f"No available buy trade found for symbol: {symbol}")
            return
        volume_string = simpledialog.askstring("Sell Stock", f"Enter number of shares to sell (Available: {buy_trade.volume}):")
        if volume_string is None or not volume_string.strip():
            return
        try:
            sell_volume = int(volume_string)
        except ValueError:
            messagebox.showerror("Input Error", "Volume must be an integer.")
            return
        if sell_volume <= 0 or sell_volume > buy_trade.volume:
            messagebox.showerror("Sell Error", f"Invalid sell volume. Must be between 1 and {buy_trade.volume}")
            return
        current_price = buy_trade.price
        trade_timestamp = time.time()
        sell_trade = Trade(trade_timestamp, symbol, current_price, sell_volume, buy_trade.original_price, "Sell")
        self.transaction_tracker.add_trade(sell_trade)
        self.portfolio_manager.add_trade(sell_trade)
        self.all_trades.append(sell_trade)
        buy_trade.reduce_volume(sell_volume)
        revenue = current_price * sell_volume
        self.wallet += revenue
        self.update_stock_history(symbol, current_price, trade_timestamp)
        self.refresh_table()
        self.update_summary()
        self.update_stock_list()

    def simulate_price_update(self, trade, max_fluctuation):
        fluctuation_factor = 1 + (self.random_generator.random() * 2 * max_fluctuation - max_fluctuation)
        return trade.price * fluctuation_factor

    def update_all_prices(self):
        for trade in self.all_trades:
            if trade.trade_type == "Buy" and trade.volume > 0:
                new_price = self.simulate_price_update(trade, 0.05)
                trade.price = new_price
        self.transaction_tracker = TransactionTracker()
        self.portfolio_manager = PortfolioManager()
        latest_prices = {}
        for trade in self.all_trades:
            self.transaction_tracker.add_trade(trade)
            self.portfolio_manager.add_trade(trade)
            if trade.trade_type == "Buy" and trade.volume > 0:
                latest_prices[trade.symbol] = trade.price
        current_time = time.time()
        for symbol, price in latest_prices.items():
            self.update_stock_history(symbol, price, current_time)
        self.refresh_table()
        self.update_summary()
        self.update_stock_list()

    def update_selected_stock(self):
        selected_items = self.trade_table.selection()
        if not selected_items:
            messagebox.showwarning("No Selection", "Please select a trade to update.")
            return
        selected_item = selected_items[0]
        table_children = self.trade_table.get_children()
        selected_index = table_children.index(selected_item)
        selected_trade = self.all_trades[selected_index]
        if selected_trade.trade_type != "Buy" or selected_trade.volume <= 0:
            messagebox.showerror("Update Error", "Selected trade is not an active buy trade.")
            return
        new_price = self.simulate_price_update(selected_trade, 0.05)
        selected_trade.price = new_price
        current_time = time.time()
        self.update_stock_history(selected_trade.symbol, new_price, current_time)
        self.transaction_tracker = TransactionTracker()
        self.portfolio_manager = PortfolioManager()
        for trade in self.all_trades:
            self.transaction_tracker.add_trade(trade)
            self.portfolio_manager.add_trade(trade)
        self.refresh_table()
        self.update_summary()
        self.update_stock_list()

    def update_summary(self):
        best_trade = self.transaction_tracker.get_best_trade()
        worst_trade = self.transaction_tracker.get_worst_trade()
        if best_trade:
            self.best_trade_label.config(text=f"Best Trade: {best_trade.symbol} | Profit: {best_trade.performance_metric():.2f}")
        else:
            self.best_trade_label.config(text="Best Trade: N/A")
        if worst_trade:
            self.worst_trade_label.config(text=f"Worst Trade: {worst_trade.symbol} | Profit: {worst_trade.performance_metric():.2f}")
        else:
            self.worst_trade_label.config(text="Worst Trade: N/A")
        self.wallet_label.config(text=f"Wallet: ${self.wallet:.2f}")

    def refresh_table(self):
        for row in self.trade_table.get_children():
            self.trade_table.delete(row)
        for trade in self.all_trades:
            self.trade_table.insert("", "end", values=(
                f"{trade.timestamp:.2f}",
                trade.symbol,
                trade.trade_type,
                f"{trade.price:.2f}",
                trade.volume,
                f"{trade.original_price:.2f}",
                f"{trade.performance_metric():.2f}"
            ))

    def update_stock_history(self, symbol, new_price, timestamp):
        period = self.CANDLE_PERIOD
        if symbol not in self.stock_history:
            self.stock_history[symbol] = [{
                "start": timestamp,
                "open": new_price,
                "high": new_price,
                "low": new_price,
                "close": new_price
            }]
        else:
            candles = self.stock_history[symbol]
            latest_candle = candles[-1]
            if timestamp < latest_candle["start"] + period:
                latest_candle["high"] = max(latest_candle["high"], new_price)
                latest_candle["low"] = min(latest_candle["low"], new_price)
                latest_candle["close"] = new_price
            else:
                candles.append({
                    "start": timestamp,
                    "open": new_price,
                    "high": new_price,
                    "low": new_price,
                    "close": new_price
                })
        if len(self.stock_history[symbol]) > 50:
            self.stock_history[symbol] = self.stock_history[symbol][-50:]

    def update_stock_list(self):
        unique_symbols = sorted({trade.symbol for trade in self.all_trades})
        self.stock_listbox.delete(0, tk.END)
        for symbol in unique_symbols:
            self.stock_listbox.insert(tk.END, symbol)

    def on_stock_select(self, event):
        selected_indices = event.widget.curselection()
        if selected_indices:
            selected_index = selected_indices[0]
            symbol = event.widget.get(selected_index)
            self.current_symbol = symbol
            self.plot_stock_history(symbol)
            if symbol in self.stock_history and self.stock_history[symbol]:
                latest_price = self.stock_history[symbol][-1]["close"]
                self.selected_stock_label.config(text=f"Selected Stock: {symbol} | Price: {latest_price:.2f}")
            else:
                self.selected_stock_label.config(text=f"Selected Stock: {symbol} | Price: N/A")

    def plot_stock_history(self, symbol):
        self.ax.clear()
        self.ax.set_title(f"{symbol} Price vs Time")
        self.ax.set_xlabel("Time")
        self.ax.set_ylabel("Price")
        if symbol not in self.stock_history or len(self.stock_history[symbol]) == 0:
            self.canvas.draw()
            return
        times = []
        prices = []
        for candle in self.stock_history[symbol]:
            time_point = datetime.fromtimestamp(candle["start"])
            times.append(time_point)
            prices.append(candle["close"])
        self.ax.plot(times, prices, marker='o', linestyle='-')
        self.figure.autofmt_xdate()
        self.canvas.draw()

if __name__ == "__main__":
    app = TradingTracker()
    app.mainloop()
